Allow AD to use primitive adjoints that provide a superset of indices

jekbradbury opened the pull request:
If a user provides a primitive adjoint for an `@differentiable` function, and AD or a `#gradient` instruction later requests the gradient of this function for a subset of the indices with respect to which the primitive adjoint provides derivatives, it would be surprising if the compiler then attempted autodiff of this function. It should instead generate gradient code that uses a subset of the indices of the primitive adjoint.

These changes:
- Add a field to the `[reverse_differentiable]` SIL attribute to mark whether the adjoint is primitive or synthetic.
- Allow synthesized gradient functions and nested adjoints to call adjoint functions that provide a superset of the required indices.
- Enable this behavior only for primitive adjoints.
- Per Richard, make `lookup` -> `lookUp` for consistency with upstream.

jekbradbury committed:
Mark primitive vs. synthetic primal/adjoint in [reverse-differentiable].

jekbradbury committed:
Allow synthesis of gradient functions using a subset of adjoint indices.

Each gradient instruction now creates a canonical (and potentially 
non-canonical) gradient function with the requested set of indices. This 
function will call the primal and adjoint with the minimal index 
superset if one exists, or synthesize a primal/adjoint with exactly the 
requested set if not. Also adds a test for this behavior.

jekbradbury committed:
Only enable superset behavior for primitively specified adjoints.

Adjoint selection during gradient and nested adjoint synthesis now
prefers the first available option of a) using an existing adjoint with
exactly the right indices, b) using the primitively specified adjoint
with the smallest available superset of indices, or c) synthesizing a
new adjoint with the exact indices.

rxwei commented on lib/SILOptimizer/Mandatory/TFDifferentiation.cpp:
```diff
@@ -940,8 +948,11 @@ class ADContext {
   StructDecl *createPrimalValueStruct(const DifferentiationTask *task);
 
-  void insertGradient(const GradientLookupKey &key, SILFunction *gradient) {
+  void insertGradient(const GradientLookupKey &key, SILFunction *gradient,
```
You don't need a separate flag to distinguish between canonical and non-canonical. Instead, use `key.second.isMaster()`.

rxwei commented on lib/SILOptimizer/Mandatory/TFDifferentiation.cpp:
```diff
@@ -4489,22 +4516,24 @@ void Differentiation::run() {
+    auto config = canonicalGrad.first.second;
+    auto *canGradFn = canonicalGrad.second;
+    auto *task = context.lookupMinimalDifferentiationTask(
+      original, config.indices);
```
Indent by 4.

rxwei reviewed (commented):
Thanks! Great patch.

It is the right direction to create a new task for each attribute at pass entry, but I don't think it handles cross-module reference cases. Please add and verify the following tests:
1. `#gradient((+) as (Float, Float) -> Float, wrt: .0)`
2. `#gradient({ (x: Float, y: Float) in x + y }, wrt: .0)`

rxwei commented on lib/ParseSIL/ParseSIL.cpp:
```diff
@@ -1126,13 +1126,23 @@ static bool parseReverseDifferentiableAttr(
   }
-  // Parse optional 'adjoint'.
+  // Parse optional '(synthesized) adjoint'.
+  bool adjointIsSynthesized;
```
Always initialize.

jekbradbury commented:
Addressed the rest of the comments. WRT cross-module references, I ran the two tests you suggested:
> #gradient((+) as (Float, Float) -> Float, wrt: .0)
> #gradient({ (x: Float, y: Float) in x + y }, wrt: .0)
The former fails with
```
error: differentiating an opaque function is not supported yet
  expectEqual(1, #gradient((+) as (Float, Float) -> Float, wrt: .0)(1, 2))
                 ^         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
while the latter already works with this patch (and I've added it to the tests). It seems like fully supporting the cross-module case will require `convention differentiable`?

rxwei commented on lib/SILOptimizer/Mandatory/TFDifferentiation.cpp:
```diff
@@ -4315,14 +4338,17 @@ static void fillCanonicalGradient(SILFunction &canGrad,
-      auto val = builder.createTupleExtract(loc, adjApply, i,
-                                            tupleSILTy.getTupleElementType(i));
-      directResults.push_back(val);
+      if (config.indices.isWrtParameter(i)) {
```
This won't work when indirect results are present. Here's an example:
```swift
@differentiable(reverse, adjoint: dfoo)
func foo<T : Differentiable>(x: Float, y: T, z: Float) -> Float

func dfoo<T : Differentiable>(x: Float, y: T, z: Float, originalResult: Float, seed: Float) -> (Float, T, Float) {
    return (..., ..., ...)
}
```

The SIL for `dfoo` is the following:
```swift
sil @dfoo : <T> (Float, @in T, Float, Float, Float) -> (Float, @out T, Float) {
bb0(%0 : $*T, %1 : $Float, %2 : $*T, %3 : $Float, %4 : $Float, %5 : $Float):
//  ^ The T result is indirectly returned
  %ret = tuple (..., ...) : $(Float, Float)
  return %ret : $(Float, Float) // <- this has no T here
}
```

Note that the `T` parameter here can only be indirectly passed. At the call site, the result tuple contains everything but indirect results. We need to map this index onto the direct result index before extracting the result. I can tackle this - if you could file a bug and leave a TODO here, that'd be great.

jekbradbury replied:
This seems like what you're looking for:
```
// TODO: unbreak this one too
@differentiable(reverse, wrt: (.0, .1), adjoint: dx_T)
func x_T<T>(_ x: Float, _ y: T) -> Float {
  if x > 1000 {
    return x
  }
  return x
}
func dx_T<T>(
    _ x: Float, _ y: T, primal: Float, seed: Float) -> (Float, T) {
  return (x, y)
}
SupersetAdjointTests.test("IndirectResults") {
  expectEqual(3, #gradient(x_T, wrt: .0)(2, Float(3)))
}
```
but it currently throws the "opaque function" error rather than reaching the issue you're describing because `findReferenceToVisibleFunction` fails (all I know is that it takes the second `return nullptr` path and not the first; everything else I try to look at crashes LLDB).

rxwei replied:
Yeah, could you file a bug for “handling indirect results when extracting a subset of primitive adjoint results”?

rxwei commented:
> while the latter already works with this patch (and I've added it to the tests). It seems like fully supporting the cross-module case will require convention differentiable?

This is to be resolved as part of the differentiation pass, apart from "generalized differentiability". Assuming we are not blocked on this, I can tackle this or next week.

rxwei reviewed (approved):
LGTM. Nice patch!

jekbradbury committed:
Address comments

jekbradbury committed:
add broken tests

jekbradbury commented:
Pushed two commented-out tests and filed https://bugs.swift.org/browse/SR-8967

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
Please fix the failing test: `simple_math.swift`.

jekbradbury committed:
Switch from !synthesized to primitive and fix logic and tests.

Marking primitive rather than synthesized makes the state of a newly 
created [reverse_differentiable] attribute (which doesn't have an 
adjoint yet but should still have had synthesized marked) less 
confusing.

jekbradbury commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
Thanks!
